src/btree/bt_cursor.c

```diff
@@ -1050,32 +1037,20 @@ __wt_btcur_search_near(WT_CURSOR_BTREE *cbt, int *exactp)
      */
     if (valid) {
         exact = cbt->compare;
+        WT_ERR(__cursor_kv_return(cbt, cbt->upd_value));
         /*
-         * Set the cursor key before the prefix search near check. If the prefix doesn't match,
-         * restore the cursor state, and continue to search for a valid key. Otherwise set the
-         * cursor value and return the valid record.
+         * This compare is needed for bounded cursors in the event that a valid key is found. The
+         * returned value of exact must reflect the comparison between the found key and the
+         * original search key, not the repositioned bounds key. This comparison ensures that is the
+         * case.
          */
-        WT_ERR(__wt_key_return(cbt));
-        if (F_ISSET(cursor, WT_CURSTD_PREFIX_SEARCH) &&
-          __wt_prefix_match(&state.key, &cursor->key) != 0)
-            __cursor_state_restore(cursor, &state);
-        else {
-            __wt_value_return(cbt, cbt->upd_value);
-            /*
-             * This compare is needed for bounded cursors in the event that a valid key is found.
-             * The returned value of exact must reflect the comparison between the found key and the
-             * original search key, not the repositioned bounds key. This comparison ensures that is
-             * the case.
-             */
-            if (WT_CURSOR_BOUNDS_SET(cursor)) {
-                if (btree->type == BTREE_ROW)
-                    WT_ERR(
-                      __wt_compare(session, btree->collator, &cursor->key, &state.key, &exact));
-                else
-                    exact = cbt->recno < state.recno ? -1 : cbt->recno == state.recno ? 0 : 1;
-            }
-            goto done;
+        if (WT_CURSOR_BOUNDS_SET(cursor)) {
+            if (btree->type == BTREE_ROW)
+                WT_ERR(__wt_compare(session, btree->collator, &cursor->key, &state.key, &exact));
+            else
+                exact = cbt->recno < state.recno ? -1 : cbt->recno == state.recno ? 0 : 1;
         }
+        goto done;
```

Comment:
We can remove the goto statement here, previously we would check if prefix matched between the current key and the prefix key. If they did not match we would force it to go into the `cursor_search_neighbouring` call. Now we can remove the goto statement, and inside connect the if statement onto L1081.

Reply:
668ac19e2